### PR TITLE
C#: Enable nullability in Semmle.Extraction project

### PR DESCRIPTION
**Pull request by calumgrant**
Adds the `<Nullable>enable</Nullable>` setting to the Semmle.Extraction and the Semmle.Autobuilder.Tests projects, and made associated fixes to remove all nullability warnings.

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction/Context.cs [hunk 1]
```diff
@@ -135,10 +159,13 @@ private void CheckEntityHasUniqueLabel(string id, ICachedEntity entity)
...
 
-        private Entity CreateNonNullEntity<Type, Entity>(ICachedEntityFactory<Type, Entity> factory, Type init) where Entity : ICachedEntity
+        public Entity CreateNonNullEntity<Type, Entity>(ICachedEntityFactory<Type, Entity> factory, Type init)
+            where Entity : ICachedEntity
         {
+            if (init is null) throw new ArgumentException("Unexpected null value", nameof(init));
+
             if (objectEntityCache.TryGetValue(init, out var cached))
-                return (Entity)cached;
+                return (Entity)cached!;
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Context.cs:168]: Why is the `!` needed? I would have expected `cached` to be non-`null` given the type `Dictionary<object, ICachedEntity>` of `objectEntityCache`.

File: csharp/extractor/Semmle.Extraction/Entities/Assembly.cs [hunk 2]
```diff
@@ -30,7 +30,7 @@ public override void Populate(TextWriter trapFile)
         {
             if (assemblyPath != null)
             {
-                trapFile.assemblies(this, File.Create(Context, assemblyPath), assembly.ToString(),
+                trapFile.assemblies(this, File.Create(Context, assemblyPath), assembly.ToString() ?? "",
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Entities/Assembly.cs:33]: Should this be `assembly.ToString()!` instead?

File: csharp/extractor/Semmle.Extraction/Entities/SourceLocation.cs [hunk 3]
```diff
@@ -1,58 +1,64 @@
+using System;
 using System.IO;
 using Microsoft.CodeAnalysis;
 
 namespace Semmle.Extraction.Entities
 {
-    public class SourceLocation : Location
+    public abstract class SourceLocation : Location {
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Entities/SourceLocation.cs:7]: Why this abstract class? It appears to only have `NonGeneratedSourceLocation` as a sub class.

File: csharp/extractor/Semmle.Extraction/Entity.cs [hunk 4]
```diff
@@ -153,7 +156,7 @@ public static void DefineLabel(this IEntity entity, TextWriter trapFile, IExtrac
             catch(Exception ex)  // lgtm[cs/catch-of-all-exceptions]
             {
                 trapFile.WriteLine("\"");
-                extractor.Message(new Message("Unhandled exception generating id", entity.ToString(), null, ex.StackTrace.ToString()));
+                extractor.Message(new Message("Unhandled exception generating id", entity.ToString() ?? "", null, ex.StackTrace));
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Entity.cs:159]: This appears to be a common pattern; would it make sense to have a `ToStringDefaultEmpty(this object o)` extension method?

File: csharp/extractor/Semmle.Extraction/Id.cs [hunk 5]
```diff
@@ -34,7 +34,12 @@ public class FreshId : IId
 
         public override string ToString() => "*";
 
-        public override bool Equals(object obj) => obj.GetType() == GetType();
+        public override bool Equals(object? obj)
+        {
+            // Expand logic to allow for nullability control flow analysis
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Id.cs:39]: What happens if you do `public override bool Equals(object obj) => obj?.GetType() == GetType();` instead?

File: csharp/extractor/Semmle.Extraction/Entities/Assembly.cs (hunk 2 above)
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Entities/Assembly.cs:33]: Well, technically `IAssemblySymbol.toString()` could return `null`, in which case we would rather use `""` than risk propagating it.

File: csharp/extractor/Semmle.Extraction/Entities/SourceLocation.cs (hunk 3 above)
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Entities/SourceLocation.cs:7]: `GeneratedLocation` also extends `SourceLocation`, and it's `abstract` for the `WriteId` and `Populate` methods.

File: csharp/extractor/Semmle.Extraction/Entity.cs (hunk 4 above)
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Entity.cs:159]: I think `?? ""` is extremely clear and succinct, and I suppose there would be cases where you want something other than `""`. Good suggestion though.

File: csharp/extractor/Semmle.Extraction/Id.cs (hunk 5 above)
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Id.cs:39]: Should be the same, so let's do that.

**Pull request merged**